### PR TITLE
Spawn with options passed in query arguments to /spawn

### PR DESCRIPTION
**Pull request by twalcari**
Currently it is only possible to pass spawn parameters via a POST-request to `/spawn` . As we want to support nbgitpuller-links, I expanded the support for spawn parameters to the GET-request of /spawn.

I also introduced an extra function `Spawner.options_from_query` to process these query parameters. However, this function is implemented to be optional: if it is not defined, it will fallback to the `Spawner.options_from_form` function.

This pull request makes it possible to create links that spawn Jupyter instances with a certain configuration, and that then use the `/user-redirect/` functionality to do something on the Jupyter notebook instance (for example: execute nbgitpuller)

Example link: https://jupyterhub.example.com/devhub/hub/spawn?project=testproj&docker_image=jupyter%2Ftensorflow-notebook%3Alatest&nprocs=1&ngpus=0&memory=2048&next=%2Fdevhub%2Fhub%2Fuser-redirect%2Fgit-pull%3Frepo%3Dhttps%253A%252F%252Fgithub.com%252FRajawat23%252FRL_Lab

### CODE REVIEW

**Comment by stv0g** [comment]
I already implemented the same function for selecting a profile in kubespawner.
But I like your implementation much more.

This would also close the following issue for us: https://github.com/jupyterhub/kubespawner/issues/189

File: jupyterhub/handlers/pages.py [hunk 1]
```diff
@@ -173,11 +173,41 @@ async def get(self, for_user=None, server_name=''):
...
+                pending_url = self._get_pending_url(user, server_name)
+                return await self._wrap_spawn_single_user(
+                    user, server_name, spawner, pending_url, options
+                )
+            except Exception as e:
+                self.log.error(
+                    "Failed to spawn single-user server with query arguments",
+                    exc_info=True,
+                )
+                error_message = str(e)
```
[the remark this answers is not in the document]
Comment (wlcrs) [reply jupyterhub/handlers/pages.py:201]: I've added the test `test_spawn_with_query_arguments_error` to test this exception branch.

**Comment by wlcrs** [comment]
> What would you expect to happen when getting a request to this URL with parameters when the server is already running with different user options? I think it's going to launch with a server that doesn't have the desired config.

Currently it will indeed try to open the `next=` parameter in the running server that hasn't got the desired config.

I could instead parse the spawn-query arguments and check if `spawner.orm_spawner.user_options` has identical values? The only drawback that I can imagine is that this won't work when some `options` are dynamically created by `Spawner.options_from_form`/`Spawner.options_from_query`. I don't think that is an issue as there are other/better places in the `Spawner` class to do such things.

File: jupyterhub/handlers/pages.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (twalcari) [reply jupyterhub/handlers/pages.py:201]: I've added the test `test_spawn_with_query_arguments_error` to test this exception branch. I'm not sure why one test are failing now in an unrelated part of the code?

**Comment by twalcari** [comment]
> What would you expect to happen when getting a request to this URL with parameters when the server is already running with different user options? I think it's going to launch with a server that doesn't have the desired config.

Currently it will indeed try to open the `next=` parameter in the running server that hasn't got the desired config.

I could instead parse the spawn-query arguments and check if `spawner.orm_spawner.user_options` has identical values? The only drawback that I can imagine is that this won't work when some `options` are dynamically created by `Spawner.options_from_form`/`Spawner.options_from_query`. I don't think that is an issue as there are other/better places in the `Spawner` class to do such things.

**Comment by minrk** [comment]
Awesome, thank you! I think we still need to think about spawning with required parameters and dealing with the case of an already running server with the wrong configuration (a longstanding issue in many ways, not really new here), but this is a straight improvement.

**Pull request merged**